### PR TITLE
wallet: random coin selection

### LINKED ISSUE #5110: Improve coin selection
Currently `lnd` coin selection always prefers the largest available utxo:

https://github.com/bottlepay/btcwallet/blob/664f77ded195410cdb1868b45c0f189b84b693be/wallet/createtx.go#L33

```golang
// Pick largest outputs first.  This is only done for compatibility with
// previous tx creation code, not because it's a good idea.
sort.Sort(sort.Reverse(byAmount(eligible)))
```

Similar logic is implemented in the sweeper and perhaps elsewhere too.

Short term, this selection logic leads to low tx fees, but over time the wallet will contain ever smaller utxos. This pushes back spending the fees to consolidate these utxos to a time where the mempool might be even fuller and bitcoin even more expensive.

One simple option is to just randomize coin selection.

Interesting background information:

https://iohk.io/en/blog/posts/2018/07/03/self-organisation-in-coin-selection/
https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf

### PR DESCRIPTION
**Pull request by joostjager**
Adds the option to use random coin selection for on-chain transfers.

Fixes #5110 

Depends on https://github.com/btcsuite/btcwallet/pull/740

### CODE REVIEW

File: lnwallet/btcwallet/config.go [hunk 1]
```diff
@@ -84,6 +84,10 @@ type Config struct {
 	// DBTimeOut specifies the timeout value to use when opening the wallet
 	// database.
 	DBTimeOut time.Duration
+
+	// CoinSelectionStrategy is the strategy that is used for selection
```
Comment (guggero) [inline lnwallet/btcwallet/config.go:88]: nit: s/selection/selecting

File: config.go [hunk 2]
```diff
@@ -500,6 +502,7 @@ func DefaultConfig() Config {
 		DB:                      lncfg.DefaultDB(),
 		registeredChains:        chainreg.NewChainRegistry(),
 		ActiveNetParams:         chainreg.BitcoinTestNetParams,
+		CoinSelectionStrategy:   "largest",
```
Comment (guggero) [inline config.go:505]: nit: should add a constant for this. Or maybe even a `String()` method on the `CoinSelectionStrategy` itself.

File: config.go [hunk 3]
```diff
@@ -271,6 +271,8 @@ type Config struct {
 
 	ResetWalletTransactions bool `long:"reset-wallet-transactions" description:"Removes all transaction history from the on-chain wallet on startup, forcing a full chain rescan starting at the wallet's birthday. Implements the same functionality as btcwallet's dropwtxmgr command. Should be set to false after successful execution to avoid rescanning on every restart of lnd."`
 
+	CoinSelectionStrategy string `long:"coin-selection-strategy" description:"The strategy to use for selecting coins." choice:"largest" choice:"random"`
```
Comment (guggero) [inline config.go:274]: Needs an entry in the `sample-lnd.conf` file.

File: chainreg/chainregistry.go [hunk 4]
```diff
@@ -105,6 +105,10 @@ type Config struct {
 	// DBTimeOut specifies the timeout value to use when opening the wallet
 	// database.
 	DBTimeOut time.Duration
+
+	// CoinSelectionStrategy is the strategy that is used for selection
```
Comment (guggero) [inline chainreg/chainregistry.go:109]: nit: s/selection/selecting

**Review by guggero — APPROVED** [review]
Very nice feature! Most of the heavy lifting is in https://github.com/btcsuite/btcwallet/pull/740, so LGTM barring the dependencies and subsequent rebases :tada:

File: config.go (hunk 3 above)
Comment (wpaulino) [reply config.go:274]: Should note that this strategy will only be used for transactions originating from the wallet and not channels, sweeps, etc. Is there any reason why you didn't include the strategy for those?

File: config.go (hunk 2 above)
Comment (joostjager) [reply config.go:505]: Added constant. Didn't want to use `String()` and surface that in the command line flags.

File: config.go (hunk 3 above)
Comment (joostjager) [reply config.go:274]: The reason that I didn't include it for those is that it is a larger and more risky refactoring. For channels and sweeps, lnd has its own coin selection logic and doesn't use the code in btcwallet. I think it is worth to move towards consolidating that, but didn't want to do it in this PR.

I think that random coin selection is mainly important for users that have high volumes of wallet transactions. Lightning channels should be long-lived, so coin selection is arguably less critical.

Updated description to make it clear that the strategy only applies to wallet transactions.

Added flag to `sample-lnd.conf`.

**Comment by joostjager** [comment]
Rebased after merge of #5047

**Review by wpaulino — APPROVED** [review]
This is ready to land once the btcwallet changes are merged into upstream.

**Comment by joostjager** [comment]
Is this PR close enough to acquire the 0.13 milestone tag?

**Comment by wpaulino** [comment]
@joostjager added. Only thing left is for @Roasbeef to approve the btcwallet PR.

**Comment by joostjager** [comment]
Rebased.

I hope you guys aren't going to let this miss the 0.13 train because of a single false-positive on the linter :pray:

**Comment by joostjager** [comment]
Rebased after merge of https://github.com/btcsuite/btcwallet/pull/740

**Comment by joostjager** [comment]
This new function could use a few lines in the release notes

**Comment by guggero** [comment]
Done. Though we usually do a sweep of all new PRs before the final release.

**Comment by joostjager** [comment]
Thanks, I thought maybe it would have been forgotten as it doesn't happen so often that new functionality is added during the rc process.

**Pull request merged**